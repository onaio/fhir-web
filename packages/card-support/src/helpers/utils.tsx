/** Function to download data to a file
 *
 * @param {string} data - data to be written to file
 * @param {string} filename - name of the file to be saved
 * @param {string} type - MIME type for the file
 */
export const downloadFile = (data: string, filename: string, type: string) => {
  const file = new Blob([data], { type });
  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
  if (window.navigator.msSaveOrOpenBlob) {
    // IE10+
    window.navigator.msSaveOrOpenBlob(file, filename);
  } else {
    // Others
    const docElement = document.createElement('a');
    const url = URL.createObjectURL(file);
    docElement.href = url;
    docElement.download = filename;
    document.body.appendChild(docElement);
    docElement.click();
    setTimeout(() => {
      document.body.removeChild(docElement);
      window.URL.revokeObjectURL(url);
    }, 0);
  }
};
