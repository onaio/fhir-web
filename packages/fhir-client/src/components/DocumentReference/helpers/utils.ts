import { get } from 'lodash';

/**
 * get object from an array, the check parameters are passed in as an obj, keys in keyValue can represent a path
 *
 * @param objArray - array obj with records of objs
 * @param keyValue - obj defining what key and values to use when matching
 * @param all - return first matched or all
 */
export function findObj<T extends object>(
  objArray?: T[],
  keyValue?: Record<keyof T | string, unknown>,
  all = false
) {
  if (objArray && keyValue) {
    const checkParameters = Object.entries(keyValue);
    const matched = objArray.filter((obj) => {
      let matches = true;
      checkParameters.forEach(([key, value]) => {
        if (get(obj, key) !== value) {
          matches = false;
        }
      });
      return matches;
    });
    if (all) {
      return matched;
    } else {
      return matched[0];
    }
  }
  if (!objArray || !keyValue) return;
}
