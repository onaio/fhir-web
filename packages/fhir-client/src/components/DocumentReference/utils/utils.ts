import { get, isEmpty, some } from 'lodash';
import { IfhirR4 } from '@smile-cdr/fhirts';
import { format } from 'date-fns';
import { dateFormat } from '../../../constants';
import FHIR from 'fhirclient';

/**
 * checks if collection is empty or if all values in the obj can be said to hold null values
 *
 * @param coll - the collection
 */
export const itemIsEmpty = <T extends object>(coll?: T) => {
  if (isEmpty(coll)) {
    return true;
  }
  return !some(coll);
};

/**
 * process top level fields of the documentReference resource
 *
 * @param docResource DocumentResource
 */
export const processTopLevelFields = (docResource: IfhirR4.IDocumentReference) => {
  const { status, type, description, securityLabel, date, id } = docResource;
  return {
    id,
    title: description,
    createdAt: date ? format(new Date(date), dateFormat) : undefined,
    status,
    documentType: {
      codeList: type?.coding,
      code: get(type, 'coding.0'),
    },
    securityCodes: {
      codeList: securityLabel,
      code: get(securityLabel, '[0].coding[0]'),
    },
  };
};

/**
 * Process context field in a documentReference
 *
 * @param docResource - the document resource
 */
export const processContextFields = (docResource: IfhirR4.IDocumentReference) => {
  const start = get(docResource, 'context.period.start');
  const end = get(docResource, 'context.period.end');
  return {
    periodStart: start ? format(new Date(start), dateFormat) : undefined,
    periodEnd: end ? format(new Date(end), dateFormat) : undefined,
    eventCoding: get(docResource, 'context.event[0].coding[0]'),
    facilityTypeCoding: get(docResource, 'context.facilityType.coding[0]'),
    practiceSettingCoding: get(docResource, 'context.practiceSetting.coding[0]'),
  };
};

/**
 * extract values from content field, will include the attachments and their formats
 *
 * @param docResource - the document resource
 */
export const processContentFields = (docResource: IfhirR4.IDocumentReference) => {
  return get(docResource, 'content').map((content) => {
    const data = get(content, 'attachment.data');
    const url = get(content, 'attachment.url');
    const absoluteUrlRegex = /^(?:[a-z]+:)?\/\//;
    const binaryUrlRegex = /Binary\/[\w-]+$/gm;
    let urlIsAbsolute, binaryUrl;
    if (url) {
      urlIsAbsolute = absoluteUrlRegex.test(url);
      const matches = Array.from(
        url.matchAll(binaryUrlRegex),
        (m) => (m as Array<string | undefined>)[0]
      );
      binaryUrl = matches[0];
    }
    const size = get(content, 'attachment.size');
    const formatCode = get(content, 'format.code');
    const formatDisplay = get(content, 'format.display');
    const formatContentType = get(content, 'attachment.contentType');

    return {
      attachment: { url, urlIsAbsolute, binaryUrl, data },
      size,
      formatCode,
      formatDisplay,
      formatContentType,
    };
  });
};

/**
 * parses docResources and extracts attachment data, filtering out those with empty string data uris
 *
 * @param docResources - documentReference resource array
 */
export const processDocumentReferences = (docResources: IfhirR4.IDocumentReference[]) => {
  return docResources.map((resource) => {
    return {
      ...processTopLevelFields(resource),
      context: processContextFields(resource),
      content: processContentFields(resource),
    };
  });
};

export type ParsedDocReference = ReturnType<typeof processDocumentReferences>;

/**
 * get the discrete and multipart of a content-type
 *
 * @param contentType - mime type
 */
export const splitContentType = (contentType: string) => {
  const splitText = contentType.split('/');
  return { discretePart: splitText[0], multipart: splitText[1] };
};

/**
 * Act as a proxy when downloading attachments so that we can customize the name.
 * This pulls attachment as blob from fhir /Binary endpoint
 *
 * @param fhirBaseURL -the fhir base url
 * @param binaryEndpoint - a Binary endpoint url path
 */
export const fetchAttachmentForDownload = async (fhirBaseURL: string, binaryEndpoint: string) => {
  return FHIR.client(fhirBaseURL)
    .request({
      url: binaryEndpoint,
    })
    .then((res) => {
      return res.blob();
    });
};
