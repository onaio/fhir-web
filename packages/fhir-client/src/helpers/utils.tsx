import { Typography, Tooltip } from 'antd';
import { Period } from '@smile-cdr/fhirts/dist/FHIR-R4/classes/period';
import { Coding } from '@smile-cdr/fhirts/dist/FHIR-R4/classes/coding';
import { Column, TableLayout } from '@opensrp/react-utils';
import type { GenericWithKey } from '@opensrp/react-utils';
import { CodeableConcept } from '@smile-cdr/fhirts/dist/FHIR-R4/classes/codeableConcept';
import React from 'react';
import { useTranslation } from '../mls';
import { get } from 'lodash';

const { Text } = Typography;

/**
 * Abstracts sort functionality for dates as strings
 *
 * @param accessor - the key name
 * @param isDate - if the string represents a date
 */
export const sorterFn =
  (accessor: string, isDate = false) =>
  (a: Record<string, unknown>, b: Record<string, unknown>) => {
    const first = get(a, accessor) as string | undefined;
    const second = get(b, accessor) as string | undefined;
    if (first === undefined) return 1;
    if (second === undefined) return -1;
    if (isDate) {
      return Date.parse(first) - Date.parse(second);
    } else {
      return first.localeCompare(second);
    }
  };

/**
 * Abstracts how to render a Fhir Period data type
 *
 * @param props - Period object
 */
export const FhirPeriod = (props: Period) => {
  const { start, end } = props;
  const { t } = useTranslation();
  return (
    <>
      <Text>{t('{{val, datetime}}', { val: new Date(start ?? '') })}</Text>-
      <Text>{t('{{val, datetime}}', { val: new Date(end ?? '') })}</Text>
    </>
  );
};

/**
 * Abstracts how Codings will be rendered
 *
 * @param root0 - props
 * @param root0.codings -  the codings
 */
export const FhirCodesTooltips = ({ codings }: { codings?: Coding[] }) => {
  return (
    <>
      {(codings ?? []).map((coding) => {
        return (
          <Tooltip key={coding.code} title={coding.system || ''}>
            <span>{coding.display}</span>
          </Tooltip>
        );
      })}
    </>
  );
};

/**
 * normalize codeCodeable concept representation
 *
 * @param concepts - codecodeableConcept
 */
export const getCodeableConcepts = (concepts?: CodeableConcept[] | CodeableConcept) => {
  const arrayConcepts = concepts ? (Array.isArray(concepts) ? concepts : [concepts]) : [];
  const rtn = arrayConcepts
    .map((codeableConcept) => {
      return codeableConcept.coding ?? [];
    })
    .flat();
  return rtn;
};

interface PatientDetailsTableProps<T, ParsedType> {
  resources: T[];
  parseResource: (resource: T) => ParsedType;
  columns: Column<ParsedType>[];
}

/**
 * Renders simple table
 *
 * @param props - component props
 */
export function PatientDetailsTable<T, ParsedType extends GenericWithKey>(
  props: PatientDetailsTableProps<T, ParsedType>
) {
  const { resources, parseResource, columns } = props;
  const tableProps = {
    datasource: resources.map(parseResource),
    columns,
  };
  return <TableLayout {...tableProps}></TableLayout>;
}
