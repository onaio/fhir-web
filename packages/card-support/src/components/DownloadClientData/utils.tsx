import Papaparse from 'papaparse';
import { OpenSRPService } from '@opensrp/server-service';
import { DownloadClientDataFormFields } from '../DownloadClientData';
import { Dispatch, SetStateAction } from 'react';
import { OPENSRP_URL_CLIENT_SEARCH, APPLICATION_CSV } from '../../constants';
import { Client } from '../../ducks/clients';
import { downloadFile } from '../../helpers/utils';
import { ParsedHierarchyNode } from '@opensrp/location-management';
import { sendErrorNotification } from '@opensrp/notifications';
import { Dictionary } from '@onaio/utils';
import lang, { Lang } from '../../lang';
/* eslint-disable @typescript-eslint/camelcase */

/** interface for user assignment response */
export interface UserAssignment {
  organizationIds: number[];
  jurisdictions: string[];
  plans: string[];
}

/** interface for CSV entries */
export interface ClientCSVEntry {
  id: string;
  dob: string;
  first_name: string;
  last_name: string;
  gender: string;
  facility_of_registration: string;
  date_of_registration: string;
  card_status: string;
  card_status_last_update: string;
}

/**
 * Handle form date change
 *
 * @param {string} range - selected date range
 * @param {Dispatch<SetStateAction<[string, string]>>} setCardOrderDate - state dispatcher
 */
export const handleCardOrderDateChange = (
  range: [string, string],
  setCardOrderDate: Dispatch<SetStateAction<[string, string]>>
): void => {
  setCardOrderDate(range);
};

/**
 * Handle download client data form submission
 *
 * @param values - sumitted form values
 * @param accessToken - OPENSRP API access token
 * @param opensrpBaseURL - OPENSRP API base URL
 * @param serviceClass - OPENSRP service class
 * @param locations location hierarchy
 * @param setSubmitting - method to set form `isSubmitting` status
 * @param langObj - tthe lang object
 */
export const submitForm = (
  values: DownloadClientDataFormFields,
  accessToken: string,
  opensrpBaseURL: string,
  serviceClass: typeof OpenSRPService,
  locations: ParsedHierarchyNode[],
  setSubmitting: (isSubmitting: boolean) => void,
  langObj: Lang = lang
): void => {
  const { clientLocation, cardStatus, cardOrderDate } = values;

  if (!clientLocation) {
    return;
  }

  setSubmitting(true);
  const startDate = cardOrderDate[0];
  const endDate = cardOrderDate[1];
  const endPoint = `${OPENSRP_URL_CLIENT_SEARCH}`;
  const serve = new serviceClass(accessToken, opensrpBaseURL, endPoint);
  let params: Dictionary = {
    locationIds: clientLocation,
  };

  if (cardStatus) {
    params = {
      ...params,
      attribute: `card_status:${cardStatus}`,
    };
  }

  serve
    .list(params)
    .then((clients: Client[]) => {
      const location = getLocationDetails(locations, clientLocation);
      const entries: ClientCSVEntry[] = clients
        .filter((client: Client) => !!client.identifiers.zeir_id)
        .filter((client: Client) => {
          const dateCreated = new Date(client.dateCreated.split('T')[0]);

          return (
            dateCreated.getTime() >= new Date(startDate).getTime() &&
            dateCreated.getTime() <= new Date(endDate).getTime()
          );
        })
        .map((client: Client) => {
          return {
            id: client.identifiers.zeir_id,
            dob: formatDDMMYYY(new Date(client.birthdate.split('T')[0])),
            first_name: client.firstName,
            last_name: client.lastName,
            gender: client.gender,
            facility_of_registration: location ? location.title : '',
            date_of_registration: formatDDMMYYY(new Date(client.dateCreated.split('T')[0])),
            card_status: client.attributes.card_status,
            card_status_last_update: formatDDMMYYY(
              new Date(client.attributes.card_status_date.split('T')[0])
            ),
          };
        });

      setSubmitting(false);

      if (entries.length) {
        createCsv(entries, buildCSVFileName(location ? location.title : '', startDate, endDate));
      } else {
        sendErrorNotification(langObj.NO_DATA_FOUND);
      }
    })
    .catch((_: Error) => {
      setSubmitting(false);
      sendErrorNotification(langObj.ERROR_OCCURRED);
    });
};

/**
 * Format a Date object as dd MM YYYY
 *
 * @param {Date} date - date object
 * @param {string} delimiter - delimiter to use to separate the different parts
 * @returns {string} formatted date
 */
export const formatDDMMYYY = (date: Date, delimiter = '/'): string => {
  //eslint-disable-next-line @typescript-eslint/restrict-plus-operands
  return `${('0' + date.getDate()).slice(-2)}${delimiter}${('0' + (date.getMonth() + 1)).slice(
    -2
  )}${delimiter}${date.getFullYear()}`;
};

/**
 * Build the CSV file name
 *
 * @param {string} location - name of the location filter
 * @param {string} startDate - start date of the range filter
 * @param {string} endDate - end date of the range filter
 * @returns {string} CSV file name
 */
export const buildCSVFileName = (location: string, startDate: string, endDate: string): string => {
  return `Children_list_${location}_${formatDDMMYYY(new Date(), '_')}_(${formatDDMMYYY(
    new Date(startDate),
    '-'
  )} - ${formatDDMMYYY(new Date(endDate), '-')})`;
};

// Create csv data for the selected jurisdiction hieracrhy
export const createCsv = (entries: ClientCSVEntry[], fileName: string): void => {
  const csv: string = Papaparse.unparse(entries, {
    header: true,
  });
  // Export csv file
  downloadFile(csv, `${fileName}.csv`, APPLICATION_CSV);
};

/**
 * Get location from the location hierarchy
 *
 * @param {ParsedHierarchyNode} locations location hierarchy
 * @param {string} locationId id of location to return
 * @returns {ParsedHierarchyNode | null} location if found, null if not found
 */
export const getLocationDetails = (
  locations: ParsedHierarchyNode[],
  locationId: string
): ParsedHierarchyNode | null => {
  let found = false;
  let i = 0;
  let location = null;

  while (!found && i < locations.length) {
    const node = locations[i];

    if (node.id && node.id === locationId) {
      found = true;
      location = node;
    } else if (node.children) {
      location = getLocationDetails(node.children, locationId);

      if (location) {
        found = true;
      }
    }

    i += 1;
  }

  return location;
};
