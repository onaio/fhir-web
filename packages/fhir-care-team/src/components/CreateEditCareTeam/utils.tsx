import { history } from '@onaio/connected-reducer-registry';
import { v4 } from 'uuid';
import FHIR from 'fhirclient';
import { sendErrorNotification, sendSuccessNotification } from '@opensrp/notifications';
import lang from '../../lang';
import { URL_CARE_TEAM } from '../../constants';
import { Dictionary } from '@onaio/utils';
import { IfhirR4 } from '@smile-cdr/fhirts';
import { FormFields } from './Form';

export const submitForm = async (
  values: FormFields,
  fhirBaseURL: string,
  id?: string,
  uuid?: string
): Promise<void> => {
  const careTeamId = uuid ? uuid : v4();
  const payload: Omit<IfhirR4.ICareTeam, 'meta'> = {
    resourceType: 'CareTeam',
    identifier: [
      {
        use: 'official',
        value: careTeamId,
      },
    ],
    id: id,
    name: values.name,
    status: values.status as IfhirR4.CareTeam.StatusEnum,
    subject: {
      reference: `Group/${values.groupsId}`,
    },
    participant: values.practitionersId?.map((id) => {
      return {
        member: {
          reference: `Practitioner/${id}`,
        },
      };
    }),
  };
  const serve = FHIR.client(fhirBaseURL);
  if (id) {
    await serve
      .update(payload)
      .then(() => sendSuccessNotification(lang.CARE_TEAMS_UPDATE_SUCCESS))
      .catch(() => sendErrorNotification(lang.ERROR_OCCURRED));
  } else {
    await serve
      .create(payload)
      .then(() => sendSuccessNotification(lang.CARE_TEAMS_ADD_SUCCESS))
      .catch(() => sendErrorNotification(lang.ERROR_OCCURRED));
  }
  history.push(URL_CARE_TEAM);
};

/** Util function to build out patient or practitioner name
 *
 * @param {object} patient - patient resource object
 * @returns {string} - returns patient or practitioner name string
 */
export function getPatientName(patient: Dictionary | undefined) {
  if (!patient) {
    return '';
  }

  let name = patient.name;
  if (!Array.isArray(name)) {
    name = [name];
  }
  name = name[0];
  if (!name) {
    return '';
  }

  const family = Array.isArray(name.family) ? name.family : [name.family];
  const given = Array.isArray(name.given) ? name.given : [name.given];
  const prefix = Array.isArray(name.prefix) ? name.prefix : [name.prefix];
  const suffix = Array.isArray(name.suffix) ? name.suffix : [name.suffix];

  return [
    prefix.map((t = '') => String(t).trim()).join(' '),
    given.map((t = '') => String(t).trim()).join(' '),
    family.map((t = '') => String(t).trim()).join(' '),
    suffix.map((t = '') => String(t).trim()).join(' '),
  ]
    .filter(Boolean)
    .join(' ');
}
