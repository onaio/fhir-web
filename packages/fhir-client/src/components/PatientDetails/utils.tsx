import React from 'react';
import type { TFunction } from '@opensrp/i18n';
import { FHIRServiceClass, ResourceDetailsProps } from '@opensrp/react-utils';
import { IResource } from '@smile-cdr/fhirts/dist/FHIR-R4/interfaces/IResource';
import { Button } from 'antd';
import { CloseOutlined } from '@ant-design/icons';
import { Link } from 'react-router-dom';
import { IImmunization } from '@smile-cdr/fhirts/dist/FHIR-R4/interfaces/IImmunization';
import { IEncounter } from '@smile-cdr/fhirts/dist/FHIR-R4/interfaces/IEncounter';
import { ICondition } from '@smile-cdr/fhirts/dist/FHIR-R4/interfaces/ICondition';
import { ITask } from '@smile-cdr/fhirts/dist/FHIR-R4/interfaces/ITask';
import { ICarePlan } from '@smile-cdr/fhirts/dist/FHIR-R4/interfaces/ICarePlan';
import {
  LIST_PATIENTS_URL,
  carePlanResourceType,
  conditionResourceType,
  encounterResourceType,
  immunizationResourceType,
  patientResourceType,
  taskResourceType,
} from '../../constants';
import { patientDetailsProps } from './ResourceSchema/Patient';
import { Dictionary } from '@onaio/utils';
import { conditionDetailsProps } from './ResourceSchema/Condition';
import { carePlanDetailsProps } from './ResourceSchema/CarePlan';
import { immunizationDetailProps } from './ResourceSchema/Immunization';
import { encounterDetailProps } from './ResourceSchema/Encounter';
import { taskDetailsProps } from './ResourceSchema/Task';

export type ResourceTypes = ICarePlan | ICondition | ITask | IImmunization | IEncounter;

/**
 * Gets default patient resource search param
 *
 * @param resourceId - Patient resource identifier
 */
export const defaultSearchParamsFactory = (resourceId: string) => {
  return { 'subject:Patient': resourceId };
};

/**
 * Build resource query
 *
 * @param fhirBaseURL - base url
 * @param resourceType - resource type
 * @param resourceId - resource id
 */
export function queryParamsFactory<T extends IResource = ResourceTypes>(
  fhirBaseURL: string,
  resourceType: string,
  resourceId: string
) {
  const resourceQueryParams = {
    queryKey: [resourceType, resourceId],
    queryFn: async () => new FHIRServiceClass<T>(fhirBaseURL, resourceType).read(resourceId),
  };
  return resourceQueryParams;
}

type PreviewDataExtractor<T> = (resource: T, t: TFunction) => ResourceDetailsProps;

/**
 * get tab table side prevew data
 *
 * @param patientId - patient resource Id
 * @param dataExtractor - Function to extract preview data
 * @param cancelHanlder - on cancel click handler
 */
export function sidePreviewDetailsExtractor<T extends IResource>(
  patientId: string,
  dataExtractor: PreviewDataExtractor<T>,
  cancelHanlder: () => void
) {
  // eslint-disable-next-line react/display-name
  return (resource: T, t: TFunction) => {
    const { resourceType, id } = resource;
    const data = dataExtractor(resource, t);
    const headerActions = (
      <Button
        data-testid="close-button"
        icon={<CloseOutlined />}
        shape="circle"
        type="text"
        size="small"
        onClick={() => cancelHanlder()}
      />
    );
    return {
      headerActions,
      footer: (
        <Link to={`${LIST_PATIENTS_URL}/${patientId}/${resourceType}/${id}`}>
          {t('View full details')}
        </Link>
      ),
      ...data,
    };
  };
}

/**
 * Gets resource detail props
 *
 * @param resourceType - resource type
 */
export const getResourceDetailsProps = (resourceType: string) => {
  const resourceDetailProps: Dictionary = {
    [carePlanResourceType]: carePlanDetailsProps,
    [patientResourceType]: patientDetailsProps,
    [conditionResourceType]: conditionDetailsProps,
    [immunizationResourceType]: immunizationDetailProps,
    [encounterResourceType]: encounterDetailProps,
    [taskResourceType]: taskDetailsProps,
  };
  const targetResourceDetailProps = resourceDetailProps[resourceType];
  return targetResourceDetailProps ?? patientDetailsProps;
};
