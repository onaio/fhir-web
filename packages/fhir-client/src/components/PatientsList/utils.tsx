import { Dictionary } from '@onaio/utils';
import get from 'lodash/get';
/* eslint-disable @typescript-eslint/no-unnecessary-condition */
/** util to extract patient name
 *
 * @param patient - patient object
 * @returns {string[]} - returns an array of name strings
 */
export function getPatientName(patient: Dictionary) {
  if (!patient) {
    return '';
  }

  let name = patient.name;
  if (!Array.isArray(name)) {
    name = [name];
  }
  name = name[0];
  if (!name) {
    return '';
  }

  const family = Array.isArray(name.family) ? name.family : [name.family];
  const given = Array.isArray(name.given) ? name.given : [name.given];
  const prefix = Array.isArray(name.prefix) ? name.prefix : [name.prefix];
  const suffix = Array.isArray(name.suffix) ? name.suffix : [name.suffix];

  return [
    prefix.map((t: string) => String(t || '').trim()).join(' '),
    given.map((t: string) => String(t || '').trim()).join(' '),
    family.map((t: string) => String(t || '').trim()).join(' '),
    suffix.map((t: string) => String(t || '').trim()).join(' '),
  ]
    .filter(Boolean)
    .join(' ');
}

/**
 * Walks thru an object (ar array) and returns the value found at the provided
 * path. This function is very simple so it intentionally does not support any
 * argument polymorphism, meaning that the path can only be a dot-separated
 * string. If the path is invalid returns undefined.
 *
 * @param {Object} obj The object (or Array) to walk through
 * @param {string} path The path (eg. "a.b.4.c")
 * @returns {*} Whatever is found in the path or undefined
 */
export function getPath(obj: Dictionary, path = '') {
  return path.split('.').reduce((out, key) => (out ? out[key] : undefined), obj);
}

/** Fumction to get observation label
 *
 * @param {Object} obj - resource object
 * @returns {string} - returns label string
 */
export function getObservationLabel(obj: Dictionary): string {
  return (
    get(obj, 'code.coding.0.display') || get(obj, 'code.text') || get(obj, 'valueQuantity.code')
  );
}

/** Fumction to get observation value quantity
 *
 * @param {Object} obj - resource object
 * @returns {string} - returns value string
 */
export function buildObservationValueString(obj: Dictionary): string {
  let quantValue = '';
  if (obj.component && Array.isArray(obj.component)) {
    obj.component.forEach((c, i) => {
      quantValue =
        quantValue +
        `${getObservationLabel(c).replace('Blood Pressure', '')}: ${
          get(c, 'valueQuantity.value') || ''
        }${get(c, 'valueQuantity.unit') || ''}${i > 0 ? '' : ', '}`;
    });
  } else {
    quantValue =
      `${get(obj, 'valueQuantity.value') || ''} ${get(obj, 'valueQuantity.unit') || ''}` || 'N/A';
  }
  return quantValue;
}
