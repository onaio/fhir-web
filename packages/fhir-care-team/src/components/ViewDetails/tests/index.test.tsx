import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { Router } from 'react-router';
import * as reactQuery from 'react-query';
import { ViewDetails } from '..';
import * as fixtures from './fixtures';
import { createBrowserHistory } from 'history';
import * as fhirCient from 'fhirclient';
import { act } from 'react-dom/test-utils';
import flushPromises from 'flush-promises';
import { createTestQueryClient } from '../../ListView/tests/utils';

const { QueryClientProvider } = reactQuery;

const testQueryClient = createTestQueryClient();

const history = createBrowserHistory();

jest.mock('@opensrp/notifications', () => ({
  __esModule: true,
  ...Object.assign({}, jest.requireActual('@opensrp/notifications')),
}));

describe('View Care Team Details', () => {
  it('works correctly', async () => {
    const fhir = jest.spyOn(fhirCient, 'client');
    fhir.mockImplementation(
      jest.fn().mockImplementation(() => {
        return {
          request: jest.fn().mockResolvedValue(fixtures.careTeam1),
        };
      })
    );
    const props = {
      careTeamId: fixtures.careTeam1.id,
      fhirBaseURL: 'https://r4.smarthealthit.org/',
    };

    const wrapper = mount(
      <Router history={history}>
        <QueryClientProvider client={testQueryClient}>
          <ViewDetails {...props} />
        </QueryClientProvider>
      </Router>
    );

    await act(async () => {
      await flushPromises();
      wrapper.update();
    });

    expect(wrapper.text()).toMatchSnapshot('nominal display');
    // att test case to capture space element props snapshot
    expect(wrapper.find('ViewDetails Space').props()).toMatchSnapshot('space element');
    wrapper.unmount();
  });

  it('displays care team, details correctly', async () => {
    const props = {
      careTeamId: fixtures.careTeam1.id,
      fhirBaseURL: 'https://r4.smarthealthit.org/',
    };
    const wrapper = mount(
      <Router history={history}>
        <QueryClientProvider client={testQueryClient}>
          <ViewDetails {...props} />
        </QueryClientProvider>
      </Router>
    );

    // check that detail view is rendered
    expect(toJson(wrapper.find('.view-details-content'))).toBeTruthy();
  });

  it('detail view without careTeamId', () => {
    const props = {
      careTeamId: '',
      fhirBaseURL: 'https://r4.smarthealthit.org/',
    };
    const wrapper = mount(
      <Router history={history}>
        <QueryClientProvider client={testQueryClient}>
          <ViewDetails {...props} />
        </QueryClientProvider>
      </Router>
    );
    expect(toJson(wrapper.find('.view-details-content'))).toMatchSnapshot('Should be null');
    wrapper.unmount();
  });

  it('Closes on clicking cancel (X) ', () => {
    const props = {
      careTeamId: fixtures.careTeam1.id,
      fhirBaseURL: 'https://r4.smarthealthit.org/',
    };
    const wrapper = mount(
      <Router history={history}>
        <QueryClientProvider client={testQueryClient}>
          <ViewDetails {...props} />
        </QueryClientProvider>
      </Router>
    );

    // simulate clicking on close button
    act(() => {
      wrapper.find('.flex-right button').simulate('click');
    });

    expect(wrapper.props().history.location.pathname).toEqual('/admin/CareTeams');
    wrapper.unmount();
  });

  it('shows broken page if fhir api is down', async () => {
    const props = {
      careTeamId: fixtures.careTeam1.id,
      fhirBaseURL: 'https://r4.smarthealthit.org/',
    };
    const reactQueryMock = jest.spyOn(reactQuery, 'useQuery');
    reactQueryMock.mockImplementation(
      () =>
        ({
          data: undefined,
          error: 'Something went wrong',
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
        } as any)
    );
    const wrapper = mount(
      <Router history={history}>
        <QueryClientProvider client={testQueryClient}>
          <ViewDetails {...props} />
        </QueryClientProvider>
      </Router>
    );

    await act(async () => {
      await flushPromises();
    });

    wrapper.update();
    /** error view */
    expect(wrapper.text()).toMatchInlineSnapshot(`"ErrorSomething went wrongGo backGo home"`);
    wrapper.unmount();
  });
});
