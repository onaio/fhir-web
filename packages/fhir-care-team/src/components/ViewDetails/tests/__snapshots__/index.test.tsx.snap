// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Care Team Details detail view without careTeamId: Should be null 1`] = `null`;

exports[`View Care Team Details works correctly: nominal display 1`] = `"NameCare Team OneIdentifier93bc9c3d-6321-41b0-9b93-1275d7114e22StatusactiveSubjectCare Team OneParticipant"`;

exports[`View Care Team Details works correctly: space element 1`] = `
Object {
  "children": Array [
    <Text
      className="display-block"
      strong={true}
    >
      Name
    </Text>,
    <Text
      className="display-block"
      type="secondary"
    >
      Care Team One
    </Text>,
    <Text
      className="display-block"
      strong={true}
    >
      Identifier
    </Text>,
    <Text
      className="display-block"
      type="secondary"
    >
      93bc9c3d-6321-41b0-9b93-1275d7114e22
    </Text>,
    <Text
      className="display-block"
      strong={true}
    >
      Status
    </Text>,
    <Text
      className="display-block"
      type="secondary"
    >
      active
    </Text>,
    <React.Fragment>
      <Text
        className="display-block"
        strong={true}
      >
        Subject
      </Text>
      <Text
        className="display-block"
        type="secondary"
      >
        Care Team One
      </Text>
    </React.Fragment>,
    <Text
      className="display-block"
      strong={true}
    >
      Participant
    </Text>,
    Array [
      <React.Fragment>
        <Text
          className="display-block"
          type="secondary"
        >
          
        </Text>
      </React.Fragment>,
      <React.Fragment>
        <Text
          className="display-block"
          type="secondary"
        >
          
        </Text>
      </React.Fragment>,
    ],
  ],
  "direction": "vertical",
}
`;
